app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt

```diff
@@ -38,8 +38,26 @@ import mozilla.components.feature.tab.collections.Tab as ComponentTab
 
 sealed class AdapterItem(@LayoutRes val viewType: Int) {
     data class TipItem(val tip: Tip) : AdapterItem(
-        ButtonTipViewHolder.LAYOUT_ID)
-    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID)
+        ButtonTipViewHolder.LAYOUT_ID
+    )
+
+    data class TopSiteList(val topSites: List<TopSite>) : AdapterItem(TopSiteViewHolder.LAYOUT_ID) {
+        override fun getChangePayload(newItem: AdapterItem): Any? {
+            (newItem as TopSiteList).let {
+                // If nothing needs to be updated visually, return a payload
+                data class TopSiteVisual(val title: String)
+                if (newItem.topSites.map { TopSiteVisual(it.title) } == this.topSites.map {
+                        TopSiteVisual(
+                            it.title
+                        )
+                    }) {
+                    return ChangePayload.NO_VISUAL_CHANGE
```

Comment:
I think it's more readable to just have if-else here. We should throw if the cast fails.